File: django-app/hyperweb/serialize.py
"""
Utilities for JSON-pickling and serialization of objects of arbitrary classes.
"""

import json
from importlib import import_module
from .errors import EncodeError, DecodeError


#####################################################################################################################################################
#####
#####  UTILITIES
#####

def classname(obj = None, cls = None):
    """Fully qualified class name of an object 'obj' or class 'cls'."""
    if cls is None: cls = obj.__class__
    name = cls.__module__ + "." + cls.__name__
    return name

def import_(fullname):
    """
    Dynamic import of a python class/function/variable given its full (dotted) package-module name.
    If no module name is present, __main__ is used.
    """
    if '.' not in fullname:
        mod, name = '__main__', fullname
        #raise Exception("Can't import an object without module/package name: %s" % path)
    else:
        mod, name = fullname.rsplit('.', 1)
    # if mod == "builtins":
    #     return getattr(globals()['__builtins__'], name)
    module = import_module(mod) #, fromlist = [mod])
    try:
        return getattr(module, name)
    except:
        raise ImportError(f"cannot import name '{name}' from '{mod}'")

def getstate(obj):
    """
    Retrieve object's state with __getstate__() or take it from __dict__.
    `obj` shall not be an instance of a standard type: int/float/list/tuple/dict/NoneType...
    """
    getstate_method = getattr(obj, '__getstate__', None)

    # call __getstate__() if present and bound ('obj' shall be an instance not a class)
    if getstate_method:
        if not hasattr(getstate_method, '__self__'): raise TypeError(f'expected an instance in getstate(), got a class')
        state = getstate_method()
        if not isinstance(state, dict):
            raise TypeError(f"The result of __getstate__() is not a dict in {obj}")
        
    # otherwise use __dict__
    else:
        state = getattr(obj, '__dict__', None)
        if state is None:
            raise TypeError(f"cannot retrieve state of an object of type <{type(obj)}>: {obj}")
    
    return state
    
def setstate(cls, state):
    """
    Create an object of a given class and set its state using __setstate__(), if present,
    or by assigning directly to __dict__ otherwise.
    """
    
    # instantiate and fill out an object of a custom class
    obj = cls()
    _setstate = getattr(obj, '__setstate__', None)
    if _setstate:
        _setstate(state)
    else:
        try:
            obj.__dict__ = dict(state)
        except:
            raise
            # raise TypeError(f"cannot assign state to an object of type <{cls}>, the state: {state}")
        
    return obj


#####################################################################################################################################################
#####
#####  JSON for arbitrary objects
#####

class JSON:
    """
    Dump & load arbitrary objects to/from JSON strings.
    Encode & decode arbitrary objects to/from JSON-compatible "state" composed of serializable types.
    Class Item is handled in a special way: only its ID is sent to output, with "(item)" flag, content omitted.
    """
    
    FLAG_ITEM  = "(item)"       # special value of ATTR_CLASS that denotes a reference to an Item
    FLAG_TYPE  = "(type)"       # special value of ATTR_CLASS that informs the value is a class rather than an instance
    FLAG_DICT  = "(dict)"       # special value of ATTR_CLASS that denotes a dict wrapper for another dict containing the reserved "@" key
    ATTR_CLASS = "@"            # special attribute appended to object state to store a class name (with package) of the object being encoded
    ATTR_STATE = "="            # special attribute to store a non-dict state of data types not handled by JSON: tuple, set, type ...
    PRIMITIVES = (bool, int, float, str, type(None))        # objects of these types are left unchanged during encoding
    
    @staticmethod
    def dump(obj, type_ = None, compact = True, **json_format):

        base_format = dict(separators = (',', ':')) if compact else {}
        json_format = {**base_format, **json_format}            # manually configured format overrides the base_format defaults
        
        state = JSON.encode(obj, type_)
        return json.dumps(state, ensure_ascii = False, **json_format)
    
    @staticmethod
    def load(dump, type_ = None):
        state = json.loads(dump)
        return JSON.decode(state, type_)
    
    @staticmethod
    def encode(obj, type_ = None):
        """
        Return a `state` that carries all the information needed for reconstruction of `obj` with decode(),
        yet it contains only JSON-compatible values and collections (possibly nested).
        Objects of custom classes are converted to dicts that store object's attributes,
        with a special attribute "@" added to hold the class name. Nested objects are encoded recursively.
        """
        t = type(obj)
        
        # retrieve object's state while checking against standard python types that need special handling
        if t in JSON.PRIMITIVES:    return obj
        if t is list:               return JSON.encode_list(obj)        # return a list, but first encode recursively all its elements

        if t is dict:
            obj = JSON.encode_dict(obj)                                 # encode_dict() always returns a dict
            #assert isinstance(obj, dict)
            if JSON.ATTR_CLASS not in obj: return obj
            return {JSON.ATTR_STATE: obj, JSON.ATTR_CLASS: JSON.FLAG_DICT}      # an "escape" wrapper is added around a dict that contains the reserved key "@"

        from hyperweb.item import Item
        if issubclass(t, Item):
            if None in obj.id: raise EncodeError(f'non-serializable Item instance with missing or incomplete ID: {obj.id}')
            id = list(obj.id)
            if t is type_: return id
            return {JSON.ATTR_STATE: id, JSON.ATTR_CLASS: JSON.FLAG_ITEM}
        
        from hyperweb.boot import registry

        if isinstance(obj, type):
            state = registry.get_path(obj)
            return {JSON.ATTR_STATE: state, JSON.ATTR_CLASS: JSON.FLAG_TYPE}
        elif t in (set, tuple):
            state = JSON.encode_list(obj)                       # warning: ordering of elements of a set in `state` is undefined and may differ between calls
        else:
            state = getstate(obj)                               # TODO: allow non-dict state from getstate()
            state = JSON.encode_dict(state)                     # recursively encode all non-standard objects inside `state`
            #assert isinstance(state, dict)
            if JSON.ATTR_CLASS in state:
                raise EncodeError(f'non-serializable object state, a reserved character "{JSON.ATTR_CLASS}" occurs as a key in the state dictionary')
            
        # if the exact class is known upfront, let's output compact state without adding "@" for class designation
        if t is type_: return state
        
        # wrap up in a dict and append class designator
        if not isinstance(state, dict):
            state = {JSON.ATTR_STATE: state}
        state[JSON.ATTR_CLASS] = registry.get_path(obj.__class__)
        
        return state
    
    @staticmethod
    def decode(state, type_ = None):
        """Reverse operation to encode(): takes an encoded JSON-serializable `state` and converts back to an object."""

        t = type(state)
        
        # decoding of a wrapped-up dict that contained a pre-existing '@' key
        if t is dict and state.get(JSON.ATTR_CLASS, None) == JSON.FLAG_DICT:
            if JSON.ATTR_STATE in state:
                state = state[JSON.ATTR_STATE]                      # `state` was a dict-wrapper around an actual dict
            return JSON.decode_dict(state)

        from hyperweb.boot import registry

        # determine the expected type `class_` of the output object
        if type_:
            if t is dict and JSON.ATTR_CLASS in state and JSON.ATTR_STATE not in state:
                raise DecodeError(f'ambiguous object state during decoding, the special key "{JSON.ATTR_CLASS}" is not needed but present: {state}')
            class_ = type_

        elif t is not dict:
            class_ = t              # an object of a standard python type must have been encoded (non-unique type, but not a dict either)

        elif JSON.ATTR_CLASS not in state:
            class_ = dict
            # raise DecodeError(f'corrupted object state during decoding, missing "{JSON.ATTR_CLASS}" key with object type designator: {state}')
        else:
            classname = state.pop(JSON.ATTR_CLASS)
            if JSON.ATTR_STATE in state:
                state_attr = state.pop(JSON.ATTR_STATE)
                if state: raise DecodeError(f'invalid serialized state, expected only {JSON.ATTR_CLASS} and {JSON.ATTR_STATE} special keys but got others: {state}')
                state = state_attr

            if classname == JSON.FLAG_ITEM:                 # decoding a reference to an Item? get a stub from the Registry
                return registry.get_item(state, load = False)
            if classname == JSON.FLAG_TYPE:                 # decoding a type (an object that represents a class or type)?
                return registry.get_class(state)            # ...get it from the Classpath

            class_ = registry.get_class(classname)          # else, decoding an INSTANCE of a type - get this type from Classpath
            
        # instantiate the output object; special handling for standard python types and Item
        if class_ in JSON.PRIMITIVES:   return state
        if class_ is list:              return JSON.decode_list(state)
        if class_ is dict:              return JSON.decode_dict(state)
        if class_ in (set, tuple):      return class_(JSON.decode_list(state))
        
        # if isinstance(class_, type):
        from .item import Item
        if issubclass(class_, Item):                        # all Item instances (stubs) must be created through the Registry
            return registry.get_item(state, load = False)
        # if issubclass(class_, type):
        #     return registry.get_class(state)

        # default object decoding via setstate()
        state = JSON.decode_dict(state)
        return setstate(class_, state)
        
        
    @staticmethod
    def encode_list(values):
        """Encode recursively all non-primitive objects inside a list."""
        return [JSON.encode(v) for v in values]
        
    @staticmethod
    def decode_list(state):
        """Decode recursively all non-primitive objects inside a list."""
        return [JSON.decode(v) for v in state]
        
    @staticmethod
    def encode_dict(state):
        """Encode recursively all non-primitive objects inside `state` dictionary."""
        for key in state:
            if type(key) is not str: raise EncodeError(f'non-serializable object state, contains a non-string key: {key}')
            # TODO: if there are any non-string keys in `state`, the entire dict must be converted to a list representation
            
        return {k: JSON.encode(v) for k, v in state.items()}

    @staticmethod
    def decode_dict(state):
        """Decode recursively all non-primitive objects inside `state` dictionary."""
        return {k: JSON.decode(v) for k, v in state.items()}

    

File: django-app/hyperweb/serialize.js
/*
Utilities for JSON-pickling and serialization of objects of arbitrary classes.
*/

import {import_module} from 'importlib';
//import {DecodeError, EncodeError} from './errors';

// function _pj_snippets(container) {
//     function in_es6(left, right) {
//         if (((right instanceof Array) || ((typeof right) === "string"))) {
//             return (right.indexOf(left) > (- 1));
//         } else {
//             if (((right instanceof Map) || (right instanceof Set) || (right instanceof WeakMap) || (right instanceof WeakSet))) {
//                 return right.has(left);
//             } else {
//                 return (left in right);
//             }
//         }
//     }
//     function set_properties(cls, props) {
//         let desc, value;
//         let _pj_a = props;
//         for (let p in _pj_a) {
//             if (_pj_a.hasOwnProperty(p)) {
//                 value = props[p];
//                 if (((((! ((value instanceof Map) || (value instanceof WeakMap))) && (value instanceof Object)) && ("get" in value)) && (value.get instanceof Function))) {
//                     desc = value;
//                 } else {
//                     desc = {"value": value, "enumerable": false, "configurable": true, "writable": true};
//                 }
//                 Object.defineProperty(cls.prototype, p, desc);
//             }
//         }
//     }
//     container["in_es6"] = in_es6;
//     container["set_properties"] = set_properties;
//     return container;
// }
// let _pj = {};
// _pj_snippets(_pj);

// function classname(obj = null, cls = null) {
//     /* Fully qualified class name of an object 'obj' or class 'cls'. */
//     let name;
//     if ((cls === null)) {
//         cls = obj.__class__;
//     }
//     name = ((cls.__module__ + ".") + cls.__name__);
//     return name;
// }
// function import_(fullname) {
//     /*
//     Dynamic import of a python class/function/variable given its full (dotted) package-module name.
//     If no module name is present, __main__ is used.
//     */
//     let mod, module, name;
//     if (!fullname.includes("."))
//         [mod, name] = ["__main__", fullname];
//     else
//         [mod, name] = fullname.rsplit(".", 1);
//
//     module = import_module(mod);
//     try {
//         return module[name];
//     } catch(e) {
//         throw new ImportError(`cannot import name '${name}' from '${mod}'`);
//     }
// }

// function getstate(obj) { return obj; }

// function getstate(obj) {
//     /*
//     Retrieve object's state with __getstate__() or take it from __dict__.
//     `obj` shall not be an instance of a standard type: int/float/list/tuple/dict/NoneType...
//     */
//     let getstate_method, state;
//     getstate_method = (obj["__getstate__"] || null);
//     if (getstate_method) {
//         if ((! ("__self__" in getstate_method))) {
//             throw new TypeError(`expected an instance in getstate(), got a class`);
//         }
//         state = getstate_method();
//         if ((! (state instanceof dict))) {
//             throw new TypeError(`The result of __getstate__() is not a dict in ${obj}`);
//         }
//     } else {
//         state = (obj["__dict__"] || null);
//         if ((state === null)) {
//             throw new TypeError(`cannot retrieve state of an object of type <${Object.getPrototypeOf(obj)}>: ${obj}`);
//         }
//     }
//     return state;
// }

// function setstate(cls, state) {
//     /*
//     Create an object of a given class and set its state using __setstate__(), if present,
//     or by assigning directly to __dict__ otherwise.
//     */
//     let _setstate, obj;
//     obj = cls();
//     _setstate = (obj["__setstate__"] || null);
//     if (_setstate) {
//         _setstate(state);
//     } else {
//         try {
//             obj.__dict__ = dict(state);
//         } catch(e) {
//             throw e;
//         }
//     }
//     return obj;
// }

class JSONx {
    /*
    Dump & load arbitrary objects to/from JSON strings.
    Encode & decode arbitrary objects to/from JSON-compatible "state" composed of serializable types.
    */
    static FLAG_ITEM  = "(item)"       // special value of ATTR_CLASS that denotes a reference to an Item
    static FLAG_TYPE  = "(type)"       // special value of ATTR_CLASS that informs the value is a class rather than an instance
    static FLAG_DICT  = "(dict)"       // special value of ATTR_CLASS that denotes a dict wrapper for another dict containing the reserved "@" key
    static ATTR_CLASS = "@"            // special attribute appended to object state to store a class name (with package) of the object being encoded
    static ATTR_STATE = "="            // special attribute to store a non-dict state of data types not handled by JSON: tuple, set, type ...

    static dump(obj, type = null) {
        let state = JSONx.encode(obj, type);
        return JSON.stringify(state);
    }
    static load(dump, type = null) {
        let state = JSON.parse(dump);
        return JSONx.decode(state, type);
    }

    static getPrototype   = (obj) => Object.getPrototypeOf(obj) ? obj !== null : null
    static getClass       = (obj) => Object.getPrototypeOf(obj).constructor ? obj !== null : null      // reading constructor from prototype is slightly safer than directly from obj
    static isPrimitiveObj = (obj) => ["number","string", "boolean"].includes(typeof obj) || obj === null || obj === undefined
    static isPrimitiveCls = (cls) => [Number, String, Boolean, null].includes(cls)
    static isArray        = (obj) => (obj && Object.getPrototypeOf(obj) === Array.prototype)
    static isDict         = (obj) => (obj && Object.getPrototypeOf(obj) === Object.prototype)
    static ofType         = (x,T) => (x && T && Object.getPrototypeOf(x) === T.prototype)      // test if x is an object of class T exactly (NOT of a subclass)
    static isClass        = (C)   => (typeof C === "function" && C.prototype !== undefined)    // test if C is a class (a constructor function with .prototype)
    static isSubclass     = (C,B) => (C === B || C.prototype instanceof B)                     // test if C is subclass of B, including C===B

    static encode(obj, type = null) {
        /*
        Return a `state` that carries all the information needed for reconstruction of `obj` with decode(),
        yet it contains only JSON-compatible values and collections (possibly nested).
        Objects of custom classes are converted to dicts that store object's attributes,
        with a special attribute "@" added to hold the class name. Nested objects are encoded recursively.
        Optional `type` constraint is a class (constructor function).
        */
        let registry = globalThis.registry
        let of_type = JSONx.ofType(obj, type)
        let state

        if (JSONx.isPrimitiveObj(obj))  return obj
        if (JSONx.isArray(obj))         return JSONx.encode_list(obj)

        if (JSONx.isDict(obj)) {
            obj = JSONx.encode_dict(obj)
            if (! JSONx.ATTR_CLASS in obj) return obj
            return {[JSONx.ATTR_STATE]: obj, [JSONx.ATTR_CLASS]: JSONx.FLAG_DICT}
        }

        let Item = registry.get_class("hyperweb.core.Item")
        if (obj instanceof Item) {
            if (!obj.has_id()) throw `non-serializable Item instance with missing or incomplete ID: ${obj.id}`
            if (of_type) return obj.id                      // `obj` is of `type_` exactly? no need to encode type info
            return {[JSONx.ATTR_STATE]: obj.id, [JSONx.ATTR_CLASS]: JSONx.FLAG_ITEM}
        }
        if (JSONx.isClass(obj)) {
            state = registry.get_path(obj)
            return {[JSONx.ATTR_STATE]: state, [JSONx.ATTR_CLASS]: JSONx.FLAG_TYPE}
        }
        else
            if (obj instanceof Set)
                state = JSONx.encode_list(Array.from(obj))
            else {
                state = JSONx.encode_dict(obj)
                if (JSONx.ATTR_CLASS in state)
                    throw `non-serializable object state, a reserved character "${JSONx.ATTR_CLASS}" occurs as a key in the state dictionary`;
            }

        // if the exact class is known upfront, let's output compact state without adding "@" for class designation
        if (of_type) return state

        // wrap up the state in a dict, if needed, and append class designator
        if (!JSONx.isDict(state))
            state = {[JSONx.ATTR_STATE]: state}

        let t = JSONx.getPrototype(obj)
        state[JSONx.ATTR_CLASS] = registry.get_path(t);

        return state;
    }

    static decode(state, type = null) {
        /*
        Reverse operation to encode(): takes an encoded JSON-serializable `state` and converts back to an object.
        Optional `type` constraint is a class (constructor function).
        */
        let registry = globalThis.registry
        let isdict = JSONx.isDict(state)
        let cls

        // decoding of a wrapped-up dict that contained a pre-existing '@' key
        if (isdict && (state[JSONx.ATTR_CLASS] === JSONx.FLAG_DICT)) {
            if (JSONx.ATTR_STATE in state)
                state = state[JSONx.ATTR_STATE]
            return JSONx.decode_dict(state)
        }

        // determine the expected class (constructor function) for the output object
        if (type) {
            if (isdict && (JSONx.ATTR_CLASS in state) && !(JSONx.ATTR_STATE in state))
                throw `ambiguous object state during decoding, the special key "${JSONx.ATTR_CLASS}" is not needed but present: ${state}`
            cls = type;
        }
        else if (!isdict)                               // `state` encodes a primitive value, or a list, or null;
            cls = JSONx.getClass(state)                 // cls=null denotes a class of null value

        else if (JSONx.ATTR_CLASS in state) {
            let classname = state[JSONx.ATTR_CLASS]
            delete state[JSONx.ATTR_CLASS]

            if (JSONx.ATTR_STATE in state) {
                let state_attr = state[JSONx.ATTR_STATE]
                if (state)
                    throw `invalid serialized state, expected only ${JSONx.ATTR_CLASS} and ${JSONx.ATTR_STATE} special keys but got others: ${state}`
                state = state_attr;
            }
            if (classname === JSONx.FLAG_ITEM)
                return registry.get_item(state);
            cls = registry.get_class(classname);
        }
        else cls = Object

        console.assert(cls !== undefined, {msg: "`cls` is undefined", state: state, type: type})

        // instantiate the output object; special handling for standard JSON types and Item
        if (JSONx.isPrimitiveCls(cls))  return state
        if (cls === Array)              return JSONx.decode_list(state)
        if (cls === Object)             return JSONx.decode_dict(state)
        if (cls === Set)                return new cls(JSONx.decode_list(state))

        let Item = registry.get_class("hyperweb.core.Item")
        if (JSONx.isSubclass(cls, Item))            // all Item instances must be created/loaded through the Registry
            return registry.get_item(state)

        let obj = JSONx.decode_dict(state)

        // return Object.create(cls, obj)
        return Object.setPrototypeOf(obj, cls)
    }

    static encode_list(values) {
        /* Encode recursively all non-primitive objects inside a list. */
        return values.map(JSONx.encode)
    }
    static decode_list(state) {
        /* Decode recursively all non-primitive objects inside a list. */
        return state.map(JSONx.decode)
    }
    static encode_dict(obj) {
        /* Encode recursively all non-primitive objects inside `state` dictionary. */
        for (const key of Object.getOwnPropertyNames(obj))
            if (typeof key !== "string")
                throw `non-serializable object state, contains a non-string key: ${key}`

        let entries = Object.entries(obj).map(([k, v]) => [k, JSONx.encode(v)])
        return Object.fromEntries(entries)

    }
    static decode_dict(state) {
        /* Decode recursively all non-primitive objects inside `state` dictionary. */
        let entries = Object.entries(state).map(([k, v]) => [k, JSONx.decode(v)])
        return Object.fromEntries(entries)
    }
}
